### PR TITLE
Create UI Component for WPT Metadata

### PR DESCRIPTION
**Pull request by KyleJu**
## Description
Add the UI component of Metadata at the bottom of the page.

- The metadata will only shown in a sub-directory.
- All triaged metadata in a sub-directory, including the children directory ones are shown.

## Review Information
Open a staging instance and go to /IndexedDB.

## Follow Up
Show the specific status when there is one - plumbing test statue from backend

### CODE REVIEW

**Comment by KyleJu** [comment]
Ready for a preliminary review

**Comment by lukebjerring** [comment]
I'm not seeing anything on https://ui-wpt-metadata-dot-wptdashboard-staging.appspot.com/results/IndexedDB

Are there some instructions missing?

File: webapp/components/wpt-metadata.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          margin-bottom: 4px;
+        }
+        .metadataNode img {
+          margin-right: 16px;
+          height: 24px;
+          width: 24px;
+        }
+      </style>
+      <div class="metadataNode">
+        <img src="/static/bug.svg" />
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]: Use `<iron-icon icon="bug-report"></iron-icon>` instead.

File: webapp/components/wpt-metadata.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        <iron-collapse id="metadata-collapsible">
+          <template is="dom-repeat" items="[[others]]" as="metadataNode">
+            <wpt-metadata-node
+              metadata-node="[[metadataNode]]"
+            ></wpt-metadata-node>
+          </template>
+        </iron-collapse>
+        <paper-button id="metadata-toggle" onclick="[[openCollapsible]]"
+          >Show more</paper-button
+        >
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]: Nit: move the closing bracket to the previous line.

File: webapp/components/wpt-metadata.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      </template>
+      <br />
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]:
```suggestion
      <br>
```
XHTML requires tags to be closed. HTML does not.

File: webapp/views/wpt-app.js [hunk 4]
```diff
@@ -237,6 +251,32 @@ class WPTApp extends PathInfo(WPTFlags(TestRunsUIBase)) {
...
+          wptMetadata.push(wptMetadataNode);
+        }
+      }
+    }
+    return wptMetadata;
+  }
+
+  _loadAllMetadata() {
+    const url = new URL('/api/metadata', window.location);
+    url.searchParams.set('products', 'safari,chrome,edge,firefox')
```
Comment (Hexcles) [inline webapp/views/wpt-app.js]: The list of products should come from the query builder.

File: webapp/views/wpt-app.js [hunk 5]
```diff
@@ -159,6 +165,11 @@ class WPTApp extends PathInfo(WPTFlags(TestRunsUIBase)) {
         type: String,
         computed: '_computePath(subroute.path)',
       },
+      metadata: Array,
```
Comment (Hexcles) [inline webapp/views/wpt-app.js]: Perhaps this, along with `displayedMetadata` and related methods, should go to `wpt-metadata`. And then `<wpt-metadata>` would take only `path` as from `wpt-app`. @lukebjerring might have more experience regarding the tradeoff here.

File: webapp/components/wpt-metadata.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+      href: {
+        type: String,
+        computed: 'computeHref(metadataNode)'
+      }
+    };
+  }
+
+  computeHref(metadataNode) {
+    const prefix = 'https://';
+    if (metadataNode.url.substr(0, prefix.length) !== prefix) {
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]:
```suggestion
    if (!metadataNode.url.startsWith(prefix)) {
```

File: webapp/components/wpt-metadata.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      others: {
+        type: Array,
+        computed: 'computeOthers(displayedMetadata)'
+      }
+    };
+  }
+
+  constructor() {
+    super();
+    this.openCollapsible = this.handleOpenCollapsible.bind(this);
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]: Do we need to declare `openCollapsible` as a property (with type function)?

File: webapp/components/wpt-metadata.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    const url = new URL('/api/metadata', window.location);
+    url.searchParams.set('products', productVal.join(','));
+    this.load(
+      window.fetch(url).then(r => r.json()).then(metadata => {
+        this.metadata = metadata;
+      })
+    );
+  }
+
+  computeDisplayedMetadata(path, metadata) {
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js]: We should have a test for this.

File: webapp/components/wpt-metadata.js (hunk 7 above)
Comment (KyleJu) [reply webapp/components/wpt-metadata.js]: openCollapsible is used strictly within this wpt-metadata custom element. I think as long as openCollapsible is not invoked and changed outside of this class, it doesn't have to be declared as a property

**Comment by KyleJu** [comment]
> The code looks good overall, but the demo doesn't seem to work -- I don't see any new stuff on the UI even after navigating to the indexeddb folder (which I believe has metadata). Could you provide a more detailed instruction of how to access the demo?

It wasnot deployed properly. I have deployed another instance and it should be good now!

**Comment by lukebjerring** [comment]
Some feedback about the UI in the demo 
- For the given bug,
  - The test name has changed
  - It's hard to see which test it applies to (in the list)
  - There's nothing indicating that it's for _Chrome_ (and for FAIL status)

Perhaps action items are
- Make the test name a link (that navigates to that test)
- Show an icon of the product when there is one
- Show the specific status when there is one

This is what I'm imagining:
image

[Test path] > [Subtest] > [Product] > [Status]
The first two could be links (actually, we don't anchor to subtest names yet. See https://github.com/web-platform-tests/wpt.fyi/issues/464)

**Comment by KyleJu** [comment]
> Some feedback about the UI in the demo
> 
> * For the given bug,
>   
>   * The test name has changed
>   * It's hard to see which test it applies to (in the list)
>   * There's nothing indicating that it's for _Chrome_ (and for FAIL status)
> 
> Perhaps action items are
> 
> * Make the test name a link (that navigates to that test)
> * Show an icon of the product when there is one
> * Show the specific status when there is one
> 
> This is what I'm imagining:
> image
> 
> [Test path] > [Subtest] > [Product] > [Status]
> The first two could be links (actually, we don't anchor to subtest names yet. See #464)

Implemented the first two suggestions in the follow-up commits - linking test name and showing icon. The third one requires some plumbing from the backend to retrieve metadata test status so I will implement it in a follow-up CL - https://github.com/web-platform-tests/wpt.fyi/projects/10#card-23632311

**Comment by lukebjerring** [comment]
On https://ui-wpt-metadata-dot-wptdashboard-staging.appspot.com/results/entries-api?label=master&label=experimental&aligned

I see the metadata, but then when I navigate to the file itself, I don't. The metadata for that file should definitely show on the single file results page too.

**Comment by lukebjerring** [comment]
"Triaged Metadata in /entries-api" doesn't read very well (is not user friendly).

How about "Relevant links for /entries-api results" ?

**Comment by lukebjerring** [comment]
Filed https://github.com/web-platform-tests/wpt.fyi/issues/1396 for detecting orphan cases (like the IndexedDB one, currently pointing to /IndexedDB/bindings-inject-key.html).

**Comment by lukebjerring** [comment]
Can we open the external link in a new tab by default (`<a target="_blank">`, I think?)

File: webapp/components/path.js [hunk 9]
```diff
@@ -57,6 +61,10 @@ const PathInfo = (superClass) => class extends superClass {
   computePathIsATestFile(path) {
     return /(\.(html|htm|py|svg|xhtml|xht|xml)(\?.*)?$)/.test(path);
   }
+
+  computePathIsRootDir(path) {
```
Comment (lukebjerring) [inline webapp/components/path.js]: Why isn't this just `path === '/'` ?

**Pull request merged**